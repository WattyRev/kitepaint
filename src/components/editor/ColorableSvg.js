import React from "react";
import PropTypes from "prop-types";
import { appliedColorsShape } from "../../containers/EditorContainer";
import Svg from "../Svg";

/**
 * Renders the provided SVG and uses the colorMap to apply a fill color to verious elements
 */
class ColorableSvg extends React.Component {
  static propTypes = {
    svg: PropTypes.string.isRequired,
    colorMap: appliedColorsShape.isRequired
  };

  componentDidMount() {
    this.applyColors();
  }

  componentDidUpdate() {
    this.applyColors();
  }

  /**
   * A DOM reference to the div containing the SVG
   */
  node = null;

  /**
   * Applies the colorMap to the rendered SVG
   */
  applyColors = () => {
    const { colorMap } = this.props;
    Object.keys(colorMap).forEach(id => {
      const color = colorMap[id].color;
      const panel = this.node.querySelectorAll(`[data-id="${id}"]`)[0];
      if (!panel) {
        return;
      }
      panel.setAttribute("fill", color);
    });
  };

  render() {
    return <Svg ref={node => (this.node = node)} svg={this.props.svg} />;
  }
}

export default ColorableSvg;
