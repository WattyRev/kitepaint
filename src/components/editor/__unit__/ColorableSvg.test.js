import React from "react";
import { shallow, mount } from "enzyme";
import ColorableSvg from "../ColorableSvg";

describe("ColorableSvg", () => {
  let defaultProps;
  beforeEach(() => {
    defaultProps = {
      svg: "<div>test</div>",
      colorMap: {}
    };
  });
  it("renders", () => {
    shallow(<ColorableSvg {...defaultProps} />);
  });
  it("Applies colors to the svg when it renders", () => {
    expect.assertions(2);
    defaultProps.svg = `<svg viewBox="0 0 1963.2 651.1">
        <polygon data-id="p1" fill="#FFFFFF" stroke="#000000" points="1769.7,48.9 1642.3,373.9 992.1,137.3 990,40.9 1069,40.6 1365.9,41.3 1549,42.7 1604.8,43.8 "/>
        <polygon data-id="p2" fill="#FFFFFF" stroke="#000000" points="1769.7,48.9 1642.3,373.9 992.1,137.3 990,40.9 1069,40.6 1365.9,41.3 1549,42.7 1604.8,43.8 "/>
      </svg>`;
    defaultProps.colorMap = {
      p2: {
        name: "red",
        color: "#ff0000"
      },
      p1: {
        name: "black",
        color: "#000000"
      }
    };
    const wrapper = mount(<ColorableSvg {...defaultProps} />)
      .find("div")
      .render();
    expect(wrapper.find("[data-id='p1']").prop("fill")).toEqual("#000000");
    expect(wrapper.find("[data-id='p2']").prop("fill")).toEqual("#ff0000");
  });
  it("Applies colors to the svg when the colorMap changes", () => {
    expect.assertions(1);
    expect.assertions(2);
    defaultProps.svg = `<svg viewBox="0 0 1963.2 651.1">
        <polygon data-id="p1" fill="#FFFFFF" stroke="#000000" points="1769.7,48.9 1642.3,373.9 992.1,137.3 990,40.9 1069,40.6 1365.9,41.3 1549,42.7 1604.8,43.8 "/>
        <polygon data-id="p2" fill="#FFFFFF" stroke="#000000" points="1769.7,48.9 1642.3,373.9 992.1,137.3 990,40.9 1069,40.6 1365.9,41.3 1549,42.7 1604.8,43.8 "/>
      </svg>`;
    defaultProps.colorMap = {
      p2: {
        name: "red",
        color: "#ff0000"
      },
      p1: {
        name: "black",
        color: "#000000"
      }
    };
    const wrapper = mount(<ColorableSvg {...defaultProps} />);
    let svgWrapper = wrapper.find("div").render();

    defaultProps.colorMap.p2 = {
      name: "gray",
      color: "#cdcdcd"
    };

    wrapper.setProps({
      colorMap: defaultProps.colorMap
    });
    svgWrapper = wrapper.find("div").render();

    expect(svgWrapper.find("[data-id='p1']").prop("fill")).toEqual("#000000");
    expect(svgWrapper.find("[data-id='p2']").prop("fill")).toEqual("#cdcdcd");
  });
});
