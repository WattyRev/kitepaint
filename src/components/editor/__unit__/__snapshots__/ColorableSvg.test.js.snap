// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorableSvg renders 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<div>test</div>",
    }
  }
/>
`;
